### ISSUE TITLE
Display Filter use causes buffer overflow

### ISSUE THREAD

**Issue by ooglek** [irontec/sngrep#225]
I'm running version 1.4.2 on a really, really old CentOS 5 box, installed from Irontec's centos binary repo. I'd try 1.4.4 but the CentOS repo hasn't been updated since 1.4.2.

Linux v1.vnum1.com 2.6.18-164.6.1.el5 #1 SMP Tue Nov 3 16:18:27 EST 2009 i686 i686 i386 GNU/Linux
CentOS release 5.3 (Final)

sngrep works fine until I hit '/' and try and search the displayed calls. I hit '/' and I get focused in the "Display Filter: " section near the top, but the next character press causes the below buffer overflow.

Let me know if this is due to me running on CentOS 5.3 or if it is a bug in sngrep!

Screen Output:
```
*** buffer overflow detected ***: /usr/bin/sngrep terminated
======= Backtrace: =========
    Method     Call State Call State Date       Time
/lib/libc.so.6(__chk_fail+0x41)[0x9d41c1]
/lib/libc.so.6(__strcpy_chk+0x43)[0x9d35e3]
/usr/bin/sngrep[0x80599c0]
/usr/bin/sngrep[0x80566b2]
/usr/bin/sngrep[0x8057dcf]
/usr/bin/sngrep[0x80521d5]
/lib/libc.so.6(__libc_start_main+0xdc)[0x903e9c]
/usr/bin/sngrep[0x804adc1]

======= Memory map: ========                                                                                                                                                                                                                                                                                                  004a9000-004c5000 r-xp 00000000 fd:00 68648976   /lib/libpcre.so.0.0.1
004c5000-004c6000 rwxp 0001c000 fd:00 68648976   /lib/libpcre.so.0.0.1
008d0000-008ea000 r-xp 00000000 fd:00 68648998   /lib/ld-2.5.so
008ea000-008eb000 r-xp 00019000 fd:00 68648998   /lib/ld-2.5.so
008eb000-008ec000 rwxp 0001a000 fd:00 68648998   /lib/ld-2.5.so
008ee000-00a2d000 r-xp 00000000 fd:00 68649000   /lib/libc-2.5.so                                                                                                                                                                                                                                                             00a2d000-00a2f000 r-xp 0013f000 fd:00 68649000   /lib/libc-2.5.so
00a2f000-00a30000 rwxp 00141000 fd:00 68649000   /lib/libc-2.5.so
00a30000-00a33000 rwxp 00a30000 00:00 0
00a35000-00a37000 r-xp 00000000 fd:00 68651362   /lib/libdl-2.5.so
00a37000-00a38000 r-xp 00001000 fd:00 68651362   /lib/libdl-2.5.so                                                                                                                                                                                                                                                            00a38000-00a39000 rwxp 00002000 fd:00 68651362   /lib/libdl-2.5.so
00a3b000-00a4e000 r-xp 00000000 fd:00 68651369   /lib/libpthread-2.5.so
00a4e000-00a4f000 r-xp 00013000 fd:00 68651369   /lib/libpthread-2.5.so
00a4f000-00a50000 rwxp 00014000 fd:00 68651369   /lib/libpthread-2.5.so
00a50000-00a52000 rwxp 00a50000 00:00 0                                                                                                                                                                                                                                                                                       00a54000-00a5b000 r-xp 00000000 fd:00 41354414   /usr/lib/libmenuw.so.5.5
00a5b000-00a5c000 rwxp 00006000 fd:00 41354414   /usr/lib/libmenuw.so.5.5
00a5e000-00a6c000 r-xp 00000000 fd:00 41355592   /usr/lib/libformw.so.5.5
00a6c000-00a6d000 rwxp 0000d000 fd:00 41355592   /usr/lib/libformw.so.5.5
00a6f000-00a72000 r-xp 00000000 fd:00 41355593   /usr/lib/libpanelw.so.5.5                                                                                                                                                                                                                                                    00a72000-00a73000 rwxp 00002000 fd:00 41355593   /usr/lib/libpanelw.so.5.5
00a7d000-00ac5000 r-xp 00000000 fd:00 41364689   /usr/lib/libncursesw.so.5.5
00ac5000-00acd000 rwxp 00048000 fd:00 41364689   /usr/lib/libncursesw.so.5.5
00acd000-00ace000 rwxp 00acd000 00:00 0                                                                                                                                                                                                                                                                                       00ad0000-00af8000 r-xp 00000000 fd:00 41355594   /usr/lib/libpcap.so.0.9.4
00af8000-00af9000 rwxp 00028000 fd:00 41355594   /usr/lib/libpcap.so.0.9.4
00aff000-00b0a000 r-xp 00000000 fd:00 68649029   /lib/libgcc_s-4.1.2-20080825.so.1
00b0a000-00b0b000 rwxp 0000a000 fd:00 68649029   /lib/libgcc_s-4.1.2-20080825.so.1
00cd1000-00cd2000 r-xp 00cd1000 00:00 0          [vdso]                                                                                                                                                                                                                                                                       08048000-0806e000 r-xp 00000000 fd:00 41354700   /usr/bin/sngrep
0806e000-0807c000 rw-p 00026000 fd:00 41354700   /usr/bin/sngrep
0807c000-0807f000 rw-p 0807c000 00:00 0
08b3e000-0986f000 rw-p 08b3e000 00:00 0          [heap]
b7200000-b725e000 rw-p b7200000 00:00 0                                                                                                                                                                                                                                                                                       b725e000-b7300000 ---p b725e000 00:00 0
b7332000-b7339000 r--s 00000000 fd:00 41451840   /usr/lib/gconv/gconv-modules.cache
b7339000-b7539000 r--p 00000000 fd:00 41368165   /usr/lib/locale/locale-archive
b7539000-b753a000 ---p b7539000 00:00 0
b753a000-b7f78000 rw-p b753a000 00:00 0                                                                                                                                                                                                                                                                                       b7f7e000-b7f80000 rw-p b7f7e000 00:00 0
bfeba000-bfecf000 rw-p bffea000 00:00 0          [stack]
Abort
Exit 134
```

**Comment by Kaian** [comment]
Hi @ooglek 

I'll try to fix this issue generating RPM packages for 1.4.4 rather than debugging what is failing in 1.4.2

It seem CentOS5 is dead and the [repositories have moved to centos vault](http://mirror.centos.org/centos/5.11/readme) so I have to update the building system to generate RPMs for centos5 from now on.

I'll drop you a line as soon as the RPMs are available.

Regards!

**Comment by linuxmaniac** [comment]
@Kaian as I mentioned you via IRC I'm having this crash too in Debian with 1.4.4 version

**Comment by Kaian** [comment]
Hmm, so this is the same problem 🤔 

It's strange this haven't shown since 1.4.2 (we use the search display often).
I'm still on holidays, so I can not access the IRC log right now, could you add the backtrace you mentioned to this issue?

Thanks!

**Comment by Kaian** [comment]
By the way, RPMs for 1.4.4 in Centos5 should be availabel for [i386](http://packages.irontec.com/centos/5/i386/) and [x86_64](http://packages.irontec.com/centos/5/x86_64)

**Comment by Kaian** [comment]
I've been digging a bit whats the code involved in the display filter and I think I've found where the problem is.

Ncurses return value length for the input field of display filter is based on the actual length of the field (when the field is empty, it returns N number of spaces). We use all available screen space to draw the input field so it's length is based on the number of columns your terminal is using. If your terminal has more than 256 columns (the size of the buffer where I copy the input field value, the crash reproduces).

https://github.com/irontec/sngrep/blob/master/src/curses/ui_call_list.c#L808
```
char dfilter[256];
strcpy(dfilter, field_buffer(info->fields[FLD_LIST_FILTER], 0));
```

You could easily test this by trying sngrep in the same way in a smaller terminal.
If this is the real problem, it is here from the beginning of sngrep 0:)

**Comment by linuxmaniac** [comment]
Yes, I just started sngrep in a smaller terminal, no problem. If I move that window to a bigger screen, the crash doesn't happen.

If I run sngrep in the bigger screen, crashes always. Confirmed

**Comment by Kaian** [comment]
Thanks for the testing @linuxmaniac :)

**Comment by Kaian** [comment]
Fixed in fd2b89526e811cc85235815197326c49446a7858

**Comment by ooglek** [comment]
Thank you!!! I'll test the 1.4.5 on my ultra-wide terminal once it is released. 

Cheers! 🥇

**Comment by linuxmaniac** [comment]
@Kaian confirmed, If I create a filter on a wide terminal when I try to remove the filter in a smaller terminal, I get a coredump, it seems the inverse of this issue

**Comment by Kaian** [comment]
Ok, I'll reopen the issue and give it a try next week, I have some theories 🤔 

Thanks for reporting!

**Comment by Kaian** [comment]
I have updated the code that handled the display filter triming so the temporal buffer has the same length that the filter field. This should prevent the crash.

But I have experienced lots of strange behaviours while testing this (using ncurses6). Reducing the terminal size actually notifies the form fields so their internal buffers are also reduced to the new available space. Increasing the terminal size sometimes makes the field larger than actual available columns, while other times keeps the original length, and in some cases after some keystrokes the buffer length changes. Other times, the buffer has space but ignores any new character added.. A total mess. 

I was unable to reproduce this myself, so some testing, at least to be sure the crash has disappeared, it's highly appreciated.

Thanks!!
